.dependency-cruiser.js

```diff
@@ -183,7 +238,140 @@ module.exports = {
           'npm-peer'
         ]
       }
-    }
+    },
+
+    /* application-specific architecture rules */
+    {
+      name: 'dependencies-of-shared',
+      severity: 'error',
+      from: {
+        path: dir('src/shared')
+      },
+      to: {
+        pathNot: [
+          dir('src/shared'),
+          dir('src/fs'),
+          ...npmDeps(
+            'node-fetch',
+            'markdown-escape',
+            'bn.js',
+          )
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-fs',
+      severity: 'error',
+      from: {
+        path: dir('src/fs')
+      },
+      to: {
+        pathNot: [
+          // TODO(SRe): is this okay?
+          dir('src/shared'), // 'shared' and 'fs' form a cycle
+          dir('src/fs'),
+          ...modules(
+            'path',
+            'fs',
+            'util',
+          ),
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-client',
+      severity: 'error',
+      from: {
+        path: dir('src/client'),
+        pathNot: [
+          file('src/client/webpack.config.js'),
+          testFilesWithinDir('src/client'),
+        ],
+      },
+      to: {
+        pathNot: [
+          dir('src/client'),
+          dir('src/shared'),
+          file('data/libraries/libraries.json'),
+          ...npmDeps(
+            '@fortawesome/fontawesome-free',
+            'clsx',
+            'easymde',
+            'query-string',
+            'react',
+            'react-alert',
+            'react-alert-template-basic',
+            'react-dom',
+            'react-loading-animation',
+            'react-portal-tooltip',
+            'react-responsive-modal',
+            'react-simplemde-editor',
+            'react-split-pane',
+            'remarkable',
+            'remarkable-react',
+            'scroll-into-view-if-needed',
+            'unicodeit',
+            'utf8',
+          ),
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-server',
+      severity: 'error',
+      from: {
+        path: dir('src/server'),
+        pathNot: [
+          dir('src/server/static'),
```

Comment:
"functions" and "static" are leafs in the sense that nothing depends on them but they should probably have their own dependency rules. Would it be better to change the structure to something like this?
```
server/
  generic/
    handlers/
    preload/
  express/
    server.ts
    routes/
  jamstack/
    functions/
    static/
```

Or maybe we should omit the "jamstack" level, since Jamstack functions are actually just plain AWS lambdas, and the "static" part isn't really Jamstack-specific either.
Anyway, then we'd have that "generic" cannot depend on anything else, and the others depend only on "generic".

Reply:
Yes, clearly separating the "jamstack" part from the node/express.js server by creating a directory for each (and for shared stuff) and writing dependency cruiser rules for dependencies between them would make a lot of sense imo.